SpellChecker.java

```diff
@@ -7,29 +7,89 @@ public static void main(String[] args) {
 		int threshold = Integer.parseInt(args[1]);
 		String[] dictionary = readDictionary("dictionary.txt");
 		String correction = spellChecker(word, threshold, dictionary);
-		System.out.println(correction);
+		System.out.println(correction); 
 	}
 
 	public static String tail(String str) {
-		// Your code goes here
+
+		String tail = str.substring(1) ;
+		
+		return tail ;
 	}
 
 	public static int levenshtein(String word1, String word2) {
-		// Your code goes here
+
+		String word1Low = word1.toLowerCase() ;
+		String word2Low = word2.toLowerCase() ;
+
+		if (word2Low.length() == 0){
+			return word1Low.length() ;
+		}
+		if (word1Low.length() == 0){
+			return word2Low.length() ;
+		}
+		if (word1Low.substring(0,1).equals(word2Low.substring(0,1))){
+			return levenshtein(tail(word1Low), tail(word2Low)) ; 
+		}
+		else{
+			int first = levenshtein(tail(word1Low),word2Low);
+			int second = levenshtein(word1Low,tail(word2Low));
+			int third = levenshtein(tail(word1Low),tail(word2Low));
+
+			int minOfThree = Math.min(Math.min(first,second ), third) ; 
+			return 1 + minOfThree ;
+			
+		}
+		
+		
 	}
 
 	public static String[] readDictionary(String fileName) {
 		String[] dictionary = new String[3000];
 
 		In in = new In(fileName);
 
-		// Your code here
+		for (int i = 0; i < 3000; i++) {
+			dictionary [i] = in.readLine() ;
+		}
 
 		return dictionary;
 	}
 
 	public static String spellChecker(String word, int threshold, String[] dictionary) {
-		// Your code goes here
+
+		String wordLow = word.toLowerCase() ;
+		String mostSimillar = dictionary[0] ;
+
+
+		for (int i = 0; i < 3000 ; i++) {
+			int distance = levenshtein(dictionary[i],wordLow);
+			if (distance <= threshold && distance < levenshtein(wordLow,mostSimillar) ){
+				mostSimillar = dictionary[i];
+			}
+
+		}
+		if(levenshtein(wordLow,mostSimillar)<= threshold){
+			return mostSimillar ; 
+		}
+		else {
+			return word ; 
+		}
```

Comment:
`levenshtein` is extremely expensive method due to it exponential growth with recursion.
You might consider to call it once for same parameters, store the result then use it whenever you need.
Here's how you could done it to avoid multiple calculations of same distance using `mostSimilarDistance` variable.
```suggestion
		String wordLow = word.toLowerCase() ;
		String mostSimillar = dictionary[0] ;
		int mostSimilarDistance = threshold + 1;

		for (int i = 0; i < 3000 ; i++) {
			int distance = levenshtein(dictionary[i],wordLow);
			if (distance <= threshold && distance < mostSimilarDistance ){
				mostSimillar = dictionary[i];
				mostSimilarDistance = distance;
			}

		}
		if(mostSimilarDistance <= threshold){
			return mostSimillar ; 
		}
			return word ; 
```